File: gcj_2016/QR_A.clj
; Google Code Jam 2016 Qualification Round
; Problem A. Counting Sheep

(use '[clojure.string :only (join split includes?)])

; generates [N N*2 N*3, ...]
(defn gen[n]
  (iterate (fn[x] (+ x n)) n))

; check if digit D is in N
(defn check[d n]
  (includes? (str n) (str d)))

; first N contains digit D
(defn find-first[d n]
  (first
    (filter (fn[x] (check d x)) (gen n))))

(defn solve[n]
  (if (= n 0)
    "INSOMNIA"
    (apply max (map (fn[d] (find-first d n)) (range 0 10)))))

(defn output-result[t ans]
  (println (str "Case #" t ": " ans)))

(def T (read-string (read-line)))
(loop [t 1]
  (when (<= t T)
    (def n (read-string (read-line)))
    (output-result t (solve n))
    (recur (inc t))))

File: gcj_2021/QR_A.cpp
// Google Code Jam 2021 Qualification Round
// Problem A. Reversort (7pts)

#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

int solve(int n, vector<int> a) {
	int ans = 0;
	for (int i = 1; i <= n - 1; ++i) {
		int pos = find(a.begin(), a.end(), i) - a.begin() + 1;
		reverse(a.begin() + i - 1, a.begin() + pos);
		ans += pos - (i - 1);
	}
	return ans;
}

int main(int argc, char *argv[]) {
	int T;
	cin >> T;
	for (int t = 1; t <= T; ++t) {
		int n;
		cin >> n;
		vector<int> a(n);
		for (int i = 0; i < n; ++i) {
			cin >> a[i];
		}
		int ans = solve(n, a);
		cout << "Case #" << t << ": " << ans << endl;
	}
	return 0;
}
